be/src/exec/stream/aggregate/stream_aggregator.h

```diff
@@ -0,0 +1,164 @@
+                                  vectorized::StreamChunkPtr* result_chunk);
+
+    // Compute agg's count for each group by keys which are used to generate different Ops.
+    Status _output_post_count_column(size_t chunk_size, vectorized::Column* agg_count_column);
+
+    // Use history(prev) output and final count to generate UPDATE_BEFORE/DELETE Ops:
+    // If prev output exists and final count > 0: generate UPDATE_BEFORE
+    // If prev output exists and final count = 0: generate DELETE
+    // @return
+    // prev_result_chunk: chunk which is generated by prev results to output
+    // pre_existence    : represents the existence for each row in the chunk.
+    Status _output_result_prev_retract_changes(size_t chunk_size, const vectorized::Columns& group_by_columns,
+                                               const vectorized::Column* agg_count_column,
+                                               vectorized::StreamChunkPtr* prev_result_chunk,
+                                               vectorized::Buffer<uint8_t>* prev_existence);
+
+    // Output incremental(post) output based on previous existence and final count:
+    // If prev output exists and final count > 0: generate UPDATE_AFTER
+    // If prev output not exists and final count > 0: generate INSERT
+    Status _output_result_post_retract_changes(size_t chunk_size, const vectorized::Columns& group_by_columns,
+                                               const vectorized::Column* agg_count_column,
+                                               const vectorized::Buffer<uint8_t>& prev_existence,
+                                               const vectorized::StreamChunkPtr& prev_result_chunk,
+                                               vectorized::StreamChunkPtr* result_chunk);
+
+    // Output results without generating retract messages.
+    Status _output_result_changes_without_retract(size_t chunk_size, const vectorized::Columns& group_by_columns,
+                                                  vectorized::StreamChunkPtr* result_chunk);
+
+private:
+    // Store buffers which can be reused in the incremental compute.
+    std::unique_ptr<MemPool> _mem_pool;
+
+    // Store group by keys to agg state map.
+    std::unique_ptr<StateTable> _result_state_table;
+    std::unique_ptr<StateTable> _intermediate_state_table;
+    // TODO: support merge into one detail table later.
+    std::vector<std::unique_ptr<StateTable>> _detail_state_tables;
+    std::vector<SlotDescriptor*> _output_slots_without_op;
+    int32_t _count_agg_idx{0};
+
+    // store all agg states
+    std::vector<std::unique_ptr<AggStateData>> _agg_func_states;
+    // output intermediate columns.
+    std::vector<int32_t> _intermediate_agg_func_ids;
+
+    std::unique_ptr<IntermediateAggGroupState> _result_agg_group;
+    std::unique_ptr<IntermediateAggGroupState> _intermediate_agg_group;
+    std::unique_ptr<DetailAggGroupState> _detail_agg_group;
+
+    std::vector<DatumRow> _non_found_keys;
```

Comment:
any usage of this field?

Reply:
yep, `_non_found_keys ` will be used in `agg_state_data` to reduce the lookups for state table.